### PR TITLE
`vtgate`: support filtering tablets by tablet-tags 

### PR DESCRIPTION
**Pull request by timvaillancourt**
## Description

This PR implements the RFC: https://github.com/vitessio/vitess/issues/15908 by:
1. Adding the `vtgate` flag: `--tablet-filter-tags StringMap` for a list of tablet-tag KVs to match
2. Considering this optional filter in the healthcheck code

This support works in addition (a secondary filter) to the mutually-exclusive `--tablet_filter` / `--keyspaces_to_watch` features

If/when there is agreement on the approach I will add documentation for the new support and flag

## Related Issue(s)

Resolves https://github.com/vitessio/vitess/issues/15908

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/vt/discovery/topology_watcher.go
Comment (timvaillancourt) [inline go/vt/discovery/topology_watcher.go:211]: Wait a second. I didn't change this, but shouldn't this be _(or condition)_:
```
if tw.tabletFilters != nil || !tw.hasTabletFiltersMatch(newVal.tablet) {
```
🤔?

**Review by deepthi — APPROVED** [review]
In general I tend to be thrifty about adding new flags, but this seems like a logical extension of the tags feature of the tablet (which already exists for whatever reason), so I'm not objecting to this.
The implementation and unit tests look fine. I'll leave it to you to investigate and fix the unit tests.

File: go/vt/discovery/healthcheck.go
Comment (frouioui) [inline go/vt/discovery/healthcheck.go:372]: Shouldn't we reset the `filters` slice here? Otherwise we keep appending the same filters over and over again for each cells.

**Review by frouioui — APPROVED** [review]
Left a comment on something i am not sure about. Otherwise, besides the failing tests LGTM.

We should probably do a single website PR that gathers the new flags introduced by this PR and https://github.com/vitessio/vitess/pull/15919.

File: go/vt/discovery/healthcheck.go
Comment (timvaillancourt) [reply go/vt/discovery/healthcheck.go:372]: @frouioui good catch! yes, this should be a reset each iteration of the loop, updated 👍

Comment (timvaillancourt) [reply go/vt/discovery/healthcheck.go:372]: @frouioui re-requested review on the fix 🙇

File: go/vt/discovery/topology_watcher.go
Comment (shlomi-noach) [inline go/vt/discovery/topology_watcher.go:411]: code styling: no need for `else`, because you've `return`ed in the `if` statement. The vitess code style prefers early returns and code unindents.

Accept this suggested change at your preference. I will wait until Monday noon/afternoon before merging as is.

```suggestion
	}
	if tablet.Tags == nil {
		return false
	}
```

**Review by shlomi-noach — APPROVED** [review]

**Comment by shlomi-noach** [comment]
Reminder for a website PR.

File: go/vt/discovery/topology_watcher.go
Comment (timvaillancourt) [reply go/vt/discovery/topology_watcher.go]: Consistency is key. Sounds good 👍

**Comment by shlomi-noach** [comment]
Will merge this PR without website changes, so as to get this into v20 before code freeze. Ideally the website PR will follow before RC1 next week.

**Pull request merged**
